### ISSUE TITLE
(coerce #(1) 'list) is slow

### ISSUE THREAD

**Issue by drmeister** [clasp-developers/clasp#225]
09:13 <stassats> (coerce #(1) 'list) is quite slow
09:19 <stassats> drmeister: in COERCE, there's an special case for LIST
09:20 <stassats> drmeister: but it's never reached, since (expand-deftype 'list) => (or cons null)
09:26 <stassats> besides, (coerce x 'list) can be transformed into (coerce-to-list x) at ct

**Comment by Bike** [comment]
I added a compiler macro for constant type. It's actually just as well that the LIST case was never hit, coz it gave a reversed list and didn't use the coerce-to-list defined elsewhere. Test:

    (let ((v (make-array 1000 :initial-element nil)))
      (time (loop repeat 1000 do (coerce v 'list))))

0.471s, 16 MB

    (let ((v (make-array 1000 :initial-element nil)))
      (time (loop repeat 1000 do (locally (declare (notinline coerce)) (coerce v 'list)))))

1.008s, 152 MB

Obviously could still be better but I don't think much of that has to do with `coerce`.

I fixed up other constant types as well. I'm a little worried about vectors, because it does `(concatenate type object)` which seems pretty roundabout. However, I tried replacing it with `(replace (make-sequence type :length (length object)) object)` and it took slightly longer, apparently because of a few checks in `replace` that `concatenate` can skip, like aliasing.

In any case it's faster than it was due to doing stuff at compile time, though not as much so. I could put a compiler macro on CONCATENATE but it mostly amounts to loop unrolling.
